<!doctype html>
<html>
  <head>
    <meta charset="utf-8">
    <meta http-equiv="X-UA-Compatible" content="chrome=1">
    <title>SynthEx by ChenMengjie</title>

    <link rel="stylesheet" href="stylesheets/styles.css">
    <link rel="stylesheet" href="stylesheets/github-dark.css">
    <script src="javascripts/scale.fix.js"></script>
    <meta name="viewport" content="width=device-width, initial-scale=1, user-scalable=no">

    <!--[if lt IE 9]>
    <script src="//html5shiv.googlecode.com/svn/trunk/html5.js"></script>
    <![endif]-->
  </head>
  <body>
    <div class="wrapper">
      <header class="without-description">
        <h1>SynthEx</h1>
        <p></p>
        <p class="view"><a href="https://github.com/ChenMengjie/SynthEx">View the Project on GitHub <small>ChenMengjie/SynthEx</small></a></p>
        <ul>
          <li><a href="https://github.com/ChenMengjie/SynthEx/zipball/master">Download <strong>ZIP File</strong></a></li>
          <li><a href="https://github.com/ChenMengjie/SynthEx/tarball/master">Download <strong>TAR Ball</strong></a></li>
          <li><a href="https://github.com/ChenMengjie/SynthEx">View On <strong>GitHub</strong></a></li>
        </ul>
      </header>
      <section>
        <h3>
<a id="synthex-a-comprehensive-suite-of-tools-for-cna-detection-and-tumor-heterogeneity-profiling-from-whole-exome-sequencing-data" class="anchor" href="#synthex-a-comprehensive-suite-of-tools-for-cna-detection-and-tumor-heterogeneity-profiling-from-whole-exome-sequencing-data" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>SynthEx: a comprehensive suite of tools for CNA detection and tumor heterogeneity profiling from whole exome sequencing data</h3>

<p><strong>SynthEx</strong> is a comprehensive suite of tools for CNA detection and tumor heterogeneity profiling. 
It is tailored to cater for the multiple characteristics of different next generation sequencing technologies. </p>

<p><strong>SynthEx</strong> is the first method aware of the sample-specific bias in targeted regions due to fold enrichment differences in tumor and normal samples. Fundamentally different from current methods, it applies different analytical strategies for single subject and large-scale cancer genomics studies. For large-scale cancer genomics studies, it corrects the technical bias due to capture efficiency by using a "synthetic normal" strategy. The "synthetic normal" samples were generated by stratifying existing normal samples (including samples outside the current study) based on capture protocols. Therefore, instead of requiring a matched pair normal sample from each tumor subject, a "synthetic normal" is used that mimics the targeted sequencing profile with the same capture efficiency level as the tumor sample. This strategy can potentially reduce both the experiment and analysis burden within a sequencing facility. </p>

<h3>
<a id="pre-processing-and-preparation" class="anchor" href="#pre-processing-and-preparation" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Pre-processing and preparation</h3>

<p><strong>SynthEx</strong> takes counts data in non-overlapping genomic bins from tumor and normal samples as input. It  has  the  option  to  input  variant  calling  results  from  tumor  sample.   Thus <strong>SynthEx</strong> relies  on other tools to generate appropriate input files. We recommend <strong>BEDtools</strong> to get the bin-level count data and freebayes for variant calling.  In addition, <strong>SynthEx</strong> calls the <strong>intersectBed</strong> function from <strong>BEDtools</strong> directly, so <strong>BEDtools</strong> should be installed.  The helper function to process VCF files is written in <strong>python</strong>. So make sure <strong>python</strong> is installed on the system if applying the script.</p>

<h3>
<a id="installation" class="anchor" href="#installation" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Installation</h3>

<p><strong>SynthEx</strong> relies on the following R packages: <strong>mclust</strong>, <strong>flsa</strong>, <strong>foreach</strong>, <strong>DNAcopy</strong>, <strong>ggplot2</strong>, <strong>gridExtra</strong>, <strong>inline</strong>, and <strong>Rcpp</strong>. Among these, <strong>DNAcopy</strong> is hosted on BioConductor and all the others are hosted on CRAN. </p>

<h3>
<a id="tutorial" class="anchor" href="#tutorial" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Tutorial</h3>

<p>A tutorial with examples illustrating the usage of <strong>SynthEx</strong> is distributed together with the package. 
After installation, the tutorial can also be invoked by <strong>SynthExUserGuide()</strong>. </p>

<h3>
<a id="author" class="anchor" href="#author" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Author</h3>

<p><strong>Mengjie Chen</strong> (UNC)</p>

<p>Bug report, comments or questions please send to <a href="mailto:mengjie@email.unc.edu">mengjie@email.unc.edu</a>.</p>
      </section>
    </div>
    <footer>
      <p>Project maintained by <a href="https://github.com/ChenMengjie">ChenMengjie</a></p>
      <p>Hosted on GitHub Pages &mdash; Theme by <a href="https://github.com/orderedlist">orderedlist</a></p>
    </footer>
    <!--[if !IE]><script>fixScale(document);</script><![endif]-->
    
  </body>
</html>
